feat(InputRating): new component

solidusite opened the pull request:
add star rating component with customizable props

<!---
☝️ PR title should follow conventional commits (https://conventionalcommits.org)
-->

### 🔗 Linked issue

Resolves #505 

<!-- If it resolves an open issue, please link the issue here. For example "Resolves #123" -->

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

A component to display and collect star ratings from users.
<img width="303" height="897" alt="image" src="https://github.com/user-attachments/assets/bbd83b40-986f-4584-bda3-a27d4c95deb7" />

<!-- Describe your changes in detail -->
<!-- Why is this change required? What problem does it solve? -->

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

solidusite committed:
feat(StarRating): add star rating component with customizable options and props

solidusite committed:
feat(StarRating): add snapshot tests for various rendering scenarios

solidusite commented:
Idk if change the name of component. Simply "rating" instead of "star-rating"?

solidusite committed:
feat(StarRating): add dark and light star rating images

caiotarifa commented:
Thanks, @solidusite.

About the component name, I’d prefer `InputRating` over `StarRating` or `Rating`.

Why:

- Functionally, this is a form control: it’s driven by v-model and supports default-value, exactly like other Input* components in Nuxt UI.  ￼
- The component isn’t truly “star” specific: the highlight icon customization (e.g. hearts), so the current name is semantically narrow.  ￼
- Keeping the Input* prefix improves discoverability and sets expectations: "this returns a value (number) through v-model".

Given it’s a new component, I think this is the best moment to align naming before release.

solidusite commented:
> Thanks, @solidusite.
> 
> 
> 
> About the component name, I’d prefer `InputRating` over `StarRating` or `Rating`.
> 
> 
> 
> Why:
> 
> 
> 
> - Functionally, this is a form control: it’s driven by v-model and supports default-value, exactly like other Input* components in Nuxt UI.  ￼
> 
> - The component isn’t truly “star” specific: the highlight icon customization (e.g. hearts), so the current name is semantically narrow.  ￼
> 
> - Keeping the Input* prefix improves discoverability and sets expectations: "this returns a value (number) through v-model".
> 
> 
> 
> Given it’s a new component, I think this is the best moment to align naming before release.

Totally agree

solidusite committed:
refactor(InputRating): renamed component to InputRating

benjamincanac commented:
We were waiting for this to be implemented in Reka UI to handle all complex logic such as form validation: https://github.com/unovue/reka-ui/pull/1968

@J-Michalek What do you think?

J-Michalek reviewed (commented):
@benjamincanac I would like to get this component done one day because there definitely are some users that want this, but I think we either need to wait for the RekaUI solution or create it ourselves based on the `RadioGroup` component.

The current implementation shown in this PR is not accessible at all and it seems there is no input so there are most likely issues with native form behavior.

solidusite commented:
I completely understand, thanks for the clarification.
I wasn’t aware of the existing RekaUI PR — if the plan is to wait for that solution, I’m totally fine with closing this PR for now.

That said, if you’d prefer to move forward sooner and have some guidance on the expected direction (for example, reworking this on top of RadioGroup to ensure proper accessibility and native form behavior), I’d be happy to iterate on the component accordingly.

benjamincanac commented:
I think it's worth iterating on your PR since you've done most of the job already to use the Reka UI RadioGroup component, it comes down to the same thing as using Rating primitives that use RadioGroup underneath.

solidusite committed:
refactor(InputRating): refactor component with RadioGroupItem, enhance component with readonly and disabled states, add focus ring customization, and improve documentation

solidusite commented:
I made some improvements @benjamincanac

### Refactoring with RadioGroupItem
- Refactored the component to use `RadioGroupItem` from `reka-ui` for better consistency with other form components
- Improved internal structure and maintainability

### Disabled vs Readonly States
Refactored the disabled/readonly logic for better clarity and visual distinction:

- **`readonly`**: Maintains normal appearance (full opacity, default cursor). Prevents user interaction while keeping the component visually unchanged. Ideal for displaying existing ratings that cannot be modified.

- **`disabled`**: Shows reduced opacity (75%) and a `not-allowed` cursor. Clearly indicates the component is temporarily unavailable or inactive. Both states prevent interaction, but `disabled` provides a stronger visual indication of unavailability.

The component now properly separates functional disabled state (which includes readonly for interaction blocking) from visual disabled state (only when explicitly disabled).

### Accessibility Improvements
- Enhanced keyboard navigation support through RadioGroupItem integration
- Improved `aria-label` attributes with descriptive text: "Rate X stars out of Y" instead of just numbers
- Better focus management and visual feedback for keyboard users
- Improved form validation integration with proper required field handling
- Focus ring is now present by default for accessibility (can be disabled via `ui` prop if needed)

### Additional Changes
- Fixed icon sizing to properly respond to size prop changes
- Removed duplicate CSS classes on icons
- Improved hover state handling when disabled
- Better visual feedback for half-star ratings
- Enhanced form integration example with validation schema

https://github.com/user-attachments/assets/8cc3c77a-598a-4525-bdb0-5636bc33ab49

J-Michalek commented on src/runtime/components/InputRating.vue:
```diff
@@ -0,0 +1,270 @@
+})
+
+const iconSizeClass = computed(() => {
+  const sizeMap: Record<string, string> = {
```
@solidusite I think this should be handled in the theme file.

J-Michalek commented on src/runtime/components/InputRating.vue:
```diff
@@ -0,0 +1,270 @@
+  return [star]
+}
+
+function getStarState(index: number): { filled: boolean, half: boolean } {
```
@solidusite Perhaps it would be better to create an array of items that already have the computed state of the star and then interate over it in the template. Right now this function gets called several time for each star.

solidusite committed:
feat(InputRating): add orientation props, update documentation and optimize stars state computation and props handling

- Remove separate computed for stars and inline logic into `starsWithState` using Array.from
- Precompute steps inside `starsWithState` and remove `getStepsForStar`
- Use `star.steps` in the template instead of calling a function per star
- Remove redundant fallback in `currentValue` (useVModel already handles `defaultValue`)
- Add optional `orientation` prop with default `horizontal` and use it in the template

solidusite committed:
docs(InputRating): remove outdated component code and streamline documentation

J-Michalek commented:
We may want to wait a little since the RekaUI PR was accepted and I think we should still prefer to have this in RekaUI if possible and just build on top of it.

solidusite commented:
That probably makes sense if it has already been merged. In the meantime, I can focus on #5781 or #5758, or feel free to point me to any other issues you think would be useful for me to work on.

benjamincanac committed:
Merge branch 'v4' into feat/star-rating-component

benjamincanac commented:
@solidusite I've merged the Reka UI v2.8 update (https://github.com/nuxt/ui/pull/5969) which includes the Rating primitives by @J-Michalek https://github.com/unovue/reka-ui/pull/1968

solidusite commented:
thx @benjamincanac , I will work on it as soon as possible

benjamincanac commented:
@solidusite Let me know if you need some help, also I added a claude skill called `contributing` recently to help ensure best practices if you want to try it out 😊

solidusite committed:
Merge branch 'v4' into feat/star-rating-component

* v4: (26 commits)
  fix(Table): improve perfs with `shallowRef` when watch deep is disabled (#6023)
  fix(ColorModeImage): add baseURL support for public paths (#6006)
  feat(Theme): new component (#4387)
  chore(readme): prevent layout shift on banner image
  chore(useEditorMenu): clean code
  docs(templates): update links
  docs(ai): add skills page
  docs: strip u prefix from component search (#6009)
  fix(useEditorMenu): rank filtered results by relevance
  chore(skills): improve & rename to `nuxt-ui`
  fix(BlogPost/ChangelogVersion): use `ImgHTMLAttributes` type for image prop (#6007)
  docs(llms): put back raw link transform after `content@3.11.1`
  fix(ChatMessages): allow message props to override role defaults (#6000)
  chore(deps): update devdependency eslint to v10 (v4) (#6019)
  chore(deps): update all non-major dependencies (v4) (#6015)
  chore(skills): add `ui` skill for usage and theming
  chore(readme): add tip on contributing skill
  feat(locale): add Belarusian language (#5972)
  chore(deps): update tiptap to ^3.19.0 (v4) (#5999)
  fix(InputMenu): prevent focus on trailing button
  ...

solidusite committed:
temp: Refactor component with Reka-ui primitive - patch needed in reka-ui

solidusite commented:
@benjamincanac in order to work we need the fix https://github.com/unovue/reka-ui/pull/2458, atm I make a patch inside the last commit.

@J-Michalek it's ok the code from the primitive?

benjamincanac committed:
Merge branch 'v4' into feat/star-rating-component

reka-ui 2.10.0 (now in v4 via #6626) ships the Rating primitive natively, so drop the temporary patches/reka-ui.patch and its pnpm-workspace.yaml entry. The component already imports RatingRoot/RatingItem/RatingItemIndicator from reka-ui; snapshots regenerated for the published build.

benjamincanac committed:
refactor(InputRating): align with current component conventions

- adopt useComponentProps theme-aware proxy + defineModel
- forward root props via local useForwardProps (not reka's) so <UTheme :props> applies
- pass raw _props to useFormField, fall back to props.x in tv() for theme defaults
- derive shared props via Pick<RatingRootProps>, use form-field id (drop manual useId)
- register missing types export and ThemeDefaults entry
- theme: semantic color tokens (text-${color} / inverted), gate transition-colors, drop orphan ring-offset
- docs: add reka-ui Rating link; data-slot on every themed element

benjamincanac committed:
refactor(InputRating): address review feedback

- theme: drop redundant orientation from defaultVariants; move hover:scale-110 to a compound variant so readonly stars no longer scale on hover
- redesign #star slot to render per-layer (empty background + inside each indicator) so the radios stay interactive when customized
- add aria-readonly for the readonly state
- inline empty-icon resolution (drop emptyStarIcon computed)
- spec: assert interaction tests unconditionally (drop if (exists()) guards)
- playground: adopt the Matrix/attrs convention
- docs: remove the non-standard Focus Ring section

benjamincanac committed:
refactor(InputRating): align with reka and polish

- inherit reka length/step instead of custom max/allowHalf
- focus: keyboard-only outline halo (has-focus-visible) matching buttons
- theme: semantic neutral color, fix xs empty-icon alignment, transition on hover scale
- redesign #star slot to render per-layer so the radios stay interactive
- add InputRating to the form example showcase
- trim playground to the Matrix convention, drop duplicate docs section + nav entry

benjamincanac committed:
docs(InputRating): document `hoverable` prop

benjamincanac commented:
@solidusite Since Reka UI 2.10 those primitives are finally exposed, I've made a few changes let me know what you think 😊

solidusite commented:
> @solidusite Since Reka UI 2.10 those primitives are finally exposed, I've made a few changes let me know what you think 😊
@benjamincanac awesome. ✅

benjamincanac committed:
docs(InputRating): add `star` icon for phosphor and tabler themes

benjamincanac committed:
refactor(InputRating): default hoverable off, rename star slot to item

- default hoverable to false (reka's hover preview doesn't reset on pointer-leave)
- rename the star slot/data-slot to item, matching reka's RatingItem
- forward modelValue via RatingRootEmits + useForwardProps instead of defineModel
- keep emptyIcon visible: render the indicator overlay transparent until active
- restore the per-radio aria-label
- docs: use filled/outline phosphor icons for the Empty Icon example
